### PR TITLE
[GlobalISel][TableGen] Take first result for multi-output instructions

### PR DESCRIPTION
**Pull request by bjope**
Previously, tblgen would reject patterns where one of its nested instructions produced more than one result. These arise when the instruction definition contains 'outs' as well as 'Defs'. This patch fixes that by always taking the first result, which is how these situations are handled in SelectionIDAG.

Differential Revision: https://reviews.llvm.org/D86617

### CODE REVIEW

File: llvm/test/TableGen/GlobalISelEmitter-multiple-output.td [hunk 1]
```diff
@@ -84,3 +86,34 @@ def : Pat<(two_in GPR32:$i1, GPR32:$i2), (TWO_INS GPR32:$i2, GPR32:$i1)>;
...
+// Test where implicit defs are added using Defs.
+
+let Defs = [R0] in
+def ImplicitDefInstr : I<(outs GPR32:$dst), (ins GPR32:$src), []>;
+def OtherInstr : I<(outs GPR32:$dst), (ins GPR32:$src), []>;
+
+def : Pat<(i32 (add i32:$src, i32:$src)),
+  (OtherInstr (ImplicitDefInstr GPR32:$src))>;
+
+// CHECK: GIM_CheckOpcode, /*MI*/0, TargetOpcode::G_ADD,
```
Comment (Pierre-vh) [inline llvm/test/TableGen/GlobalISelEmitter-multiple-output.td:100]: this will definitely fail because we now use variable-length encodings (`GIMT_Encode` macros)

File: llvm/utils/TableGen/GlobalISelEmitter.cpp [hunk 2]
```diff
@@ -288,11 +288,21 @@ static std::string getMangledRootDefName(StringRef DefOperandName) {
 
 //===- GlobalISelEmitter class --------------------------------------------===//
 
-static Expected<LLTCodeGen> getInstResultType(const TreePatternNode *Dst) {
+static Expected<LLTCodeGen> getInstResultType(const TreePatternNode *Dst,
+                                              const CodeGenTarget &Target) {
+  // While allowing more than one output (both implicit and explicit defs)
```
Comment (Pierre-vh) [inline llvm/utils/TableGen/GlobalISelEmitter.cpp:293]: nit: phrasing: `While we allow`?

File: llvm/utils/TableGen/GlobalISelEmitter.cpp [hunk 3]
```diff
@@ -288,11 +288,21 @@ static std::string getMangledRootDefName(StringRef DefOperandName) {
...
+  // below, we only expect one explicit def here.
+  assert(Dst->getOperator()->isSubClassOf("Instruction"));
+  CodeGenInstruction &InstInfo = Target.getInstruction(Dst->getOperator());
+  if (InstInfo.Operands.NumDefs != 1)
+    return failedImport("Dst pattern child is not having single result");
```
Comment (Pierre-vh) [inline llvm/utils/TableGen/GlobalISelEmitter.cpp:298]: nit: phrasing: `has more than one result`?

File: llvm/utils/TableGen/GlobalISelEmitter.cpp [hunk 4]
```diff
@@ -288,11 +288,21 @@ static std::string getMangledRootDefName(StringRef DefOperandName) {
...
 //===- GlobalISelEmitter class --------------------------------------------===//
 
-static Expected<LLTCodeGen> getInstResultType(const TreePatternNode *Dst) {
+static Expected<LLTCodeGen> getInstResultType(const TreePatternNode *Dst,
+                                              const CodeGenTarget &Target) {
+  // While allowing more than one output (both implicit and explicit defs)
+  // below, we only expect one explicit def here.
+  assert(Dst->getOperator()->isSubClassOf("Instruction"));
+  CodeGenInstruction &InstInfo = Target.getInstruction(Dst->getOperator());
+  if (InstInfo.Operands.NumDefs != 1)
```
Comment (Pierre-vh) [inline llvm/utils/TableGen/GlobalISelEmitter.cpp:297]: why do we need to enforce only one def?
for `v_mad_u64_u32` I'd need to allow 2 defs I think

File: llvm/test/CodeGen/AMDGPU/GlobalISel/inst-select-fneg.mir [hunk 5]
```diff
@@ -1023,44 +1027,48 @@ body: |
...
     ; GFX9: liveins: $sgpr0_sgpr1
     ; GFX9-NEXT: {{  $}}
     ; GFX9-NEXT: [[COPY:%[0-9]+]]:sreg_64 = COPY $sgpr0_sgpr1
-    ; GFX9-NEXT: [[COPY1:%[0-9]+]]:sreg_32 = COPY [[COPY]].sub0
-    ; GFX9-NEXT: [[COPY2:%[0-9]+]]:sreg_32 = COPY [[COPY]].sub1
     ; GFX9-NEXT: [[S_MOV_B32_:%[0-9]+]]:sreg_32 = S_MOV_B32 2147483648
-    ; GFX9-NEXT: [[S_OR_B32_:%[0-9]+]]:sreg_32 = S_OR_B32 [[COPY2]], [[S_MOV_B32_]], implicit-def dead $scc
-    ; GFX9-NEXT: [[REG_SEQUENCE:%[0-9]+]]:sreg_64 = REG_SEQUENCE [[COPY1]], %subreg.sub0, [[S_OR_B32_]], %subreg.sub1
+    ; GFX9-NEXT: [[COPY1:%[0-9]+]]:sreg_32_xm0 = COPY [[COPY]].sub1
+    ; GFX9-NEXT: [[S_OR_B32_:%[0-9]+]]:sreg_32 = S_OR_B32 [[COPY1]], [[S_MOV_B32_]], implicit-def dead $scc
```
Comment (Pierre-vh) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/inst-select-fneg.mir:1056]: this is fine, I'm just curious why this change causes the code to change in such a way

File: llvm/utils/TableGen/GlobalISelEmitter.cpp (hunk 3 above)
Comment (bjope) [reply llvm/utils/TableGen/GlobalISelEmitter.cpp:298]: Well, it could be zero as well when failing here. So it is "not one", which isn't same thing as "more than one".

File: llvm/utils/TableGen/GlobalISelEmitter.cpp (hunk 4 above)
Comment (bjope) [reply llvm/utils/TableGen/GlobalISelEmitter.cpp:297]: If I remember correctly this check was added to make the patch less general than the first proposal in Phabricator by Gabriel. So I wanted to narrow it down to the specific scenario that we needed downstream (i.e. allowing additional implicit defs beside the one-and-only explicit def.

With this PR we still report failure when the number of explicit defs aren't equal to one , just like in the baseline version. But it allows importing pattern in situation when we find additional implicit defs.

File: llvm/test/TableGen/GlobalISelEmitter-multiple-output.td (hunk 1 above)
Comment (bjope) [reply llvm/test/TableGen/GlobalISelEmitter-multiple-output.td:100]: Yes, I noticed that it failed and needs an update. I'll push a fixup for this.

File: llvm/utils/TableGen/GlobalISelEmitter.cpp (hunk 4 above)
Comment (Pierre-vh) [reply llvm/utils/TableGen/GlobalISelEmitter.cpp:298]: Is it a lot of work to allow any `NumDefs` ? Is it just updating tests and fixing potential regressions?
If so that's alright and I can do it.

**Review by Pierre-vh — APPROVED** [review]
LGTM but wait for @arsenm to comment, I'm really wondering about those instselect test changes

**Comment by s-barannikov** [comment]
@bjope 
> So I wanted to narrow it down to the specific scenario that we needed downstream (i.e. allowing additional implicit defs beside the one-and-only explicit def.

I've been having the same issue, but this patch does not help with it. Could you show how do you declare such instructions and patterns?

Here is how I do it.

An instruction is defined like this (I've omitted irrelevant pieces):
```
def ADDrf : Instruction {
  let OutOperandList = (outs GR:$rd);
  let InOperandList = (ins GR:$rs2, GR:$rs1);
  let Defs = [PSWR];
}
```

SDagISel part:
```
def NM_SDTBinOpWithFlagsOut : SDTypeProfile<2, 2, [
  SDTCisInt<0>,         // result
  SDTCisVT<1, FlagsVT>, // out flags
  SDTCisSameAs<2, 0>,   // lhs
  SDTCisSameAs<3, 0>    // rhs
]>;

def nm_add : SDNode<"NMISD::ADD", NM_SDTBinOpWithFlagsOut, [SDNPCommutative]>;

def : Pat<(nm_add i32:$lhs, i32:$rhs), (ADDrf i32:$lhs, i32:$rhs)>;
```

GISel part:
```
def G_NM_ADD : NMGenericInst {
  let OutOperandList = (outs type0:$dst, type1:$flags_out);
  let InOperandList = (ins type0:$lhs, type0:$rhs);
}

def : GINodeEquiv<G_NM_ADD, nm_add>;
```

SDagISel is able to use the pattern, bug -gen-global-isel reports that the pattern cannot be imported:
```
NMInstrPatterns.td:231:1: warning: Skipped pattern: Src pattern result has more defs than dst MI (2 def(s) vs 1 def(s))
def : Pat<(nm_add i32:$lhs, i32:$rhs), (ADDrf i32:$lhs, i32:$rhs)>;
```

I tried to fix the issue too, but got lost figuring out how SDagISel deals with it.

**Comment by bjope** [comment]
@s-barannikov : First of all this patch is about handling one explicit def plus extra implicit defs.
Typical example is the test case I added looking like this:
```
let Defs = [R0] in
def ImplicitDefInstr : I<(outs GPR32:$dst), (ins GPR32:$src), []>;
def OtherInstr : I<(outs GPR32:$dst), (ins GPR32:$src), []>;

def : Pat<(i32 (add i32:$src, i32:$src)),
  (OtherInstr (ImplicitDefInstr GPR32:$src))>;
```

Your example involves multiple explicit outs. The ADDrf instruction is only having one explicit def, but nm_add is having two outs. So I don't really know how that would work either with SDagISel. Afaik you can't have patterns that produce more than one outs and match that in the selection DAG.

Are you sure that the

> def : Pat<(nm_add i32:$lhs, i32:$rhs), (ADDrf i32:$lhs, i32:$rhs)>;

pattern really works as intended with SDagISel?
(Apparently tablegen has accepted the pattern, but I wonder if it ever is matched and if you really get the MIR that you expect in such cases.)

**Comment by s-barannikov** [comment]
@bjope 

Thanks for the example, I didn't notice the added test. It indeed looks like that you're addressing a different issue than mine. Unfortunately :slightly_smiling_face: 

> Are you sure that the pattern really works as intended with SDagISel?

Yes, I'm pretty sure. I was surprised that it works, but it does, and is very handy. I'm selecting all ALU instructions this way with no custom C++ code.
The support is built around `CodeGenInstruction::HasOneImplicitDefWithKnownVT()` somehow. I tried to port this approach to GISel but couldn't finish the job quickly and switched to another task.

**Comment by bjope** [comment]
> LGTM but wait for @arsenm to comment, I'm really wondering about those instselect test changes

I think that it somewhat is related to this comment:
```
bool AMDGPUInstructionSelector::selectG_FNEG(MachineInstr &MI) const {
  // Only manually handle the f64 SGPR case.
  //
  // FIXME: This is a workaround for 2.5 different tablegen problems. Because
  // the bit ops theoretically have a second result due to the implicit def of
  // SCC, the GlobalISelEmitter is overly conservative and rejects it. Fixing
  // that is easy by disabling the check. The result works, but uses a
  // nonsensical sreg32orlds_and_sreg_1 regclass.
  //
  // The DAG emitter is more problematic, and incorrectly adds both S_XOR_B32 to
  // the variadic REG_SEQUENCE operands.
```
And for example AMDGPUInstructionSelector::selectG_FABS refer to also being a fix for the same problem.

I guess that this patch fixes the problem with the implicit def. But I don't know enough about AMDGPU to understand the consequence here. And if fxing the tablegen problem makes the result worse here or not. Neither if the FIXME should be updated or removed (nor if the workarounds in AMDGPUInstructionSelector can be removed after this patch).

@Pierre-vh : I agree that it would be nice with some kind of feedback from @arsenm as well.

**Comment by bjope** [comment]
FYI: I just force pushed an update to resolve merge conflicts (to keep this patch up to date with main trunk).

**Comment by Pierre-vh** [comment]
I suspect the difference is due to the fact that TG now handles the pattern instead of the C++ code. Can you verify by checking if it still works if you remove selectG_FABS/G_FNEG?

**Comment by bjope** [comment]
@Pierre-vh : Still nothing from @arsenm , so can I push this given your earlier LGTM?

I still think someone from AMDGPU could look in to cleaning up selectG_FABS/G_FNEG etc. after on this patch. I don't feel comfortable doing that.

But it is really annoying that this patch has been stalled since August 2020, even if AMD people have shown interest in getting it merged. And there has been LGTM:s with the only catch being that @arsenm must approve it, but @arsenm hasn't commented on it since October 2020.

**Comment by Pierre-vh** [comment]
Just push it yes, if Matt has comments and some AMDGPU changes are needed I'll do them.

File: llvm/test/CodeGen/AMDGPU/GlobalISel/inst-select-fabs.mir [hunk 6]
```diff
@@ -426,44 +426,48 @@ body: |
...
     ; SI-NEXT: [[COPY:%[0-9]+]]:sreg_64 = COPY $sgpr0_sgpr1
-    ; SI-NEXT: [[COPY1:%[0-9]+]]:sreg_32 = COPY [[COPY]].sub0
-    ; SI-NEXT: [[COPY2:%[0-9]+]]:sreg_32 = COPY [[COPY]].sub1
     ; SI-NEXT: [[S_MOV_B32_:%[0-9]+]]:sreg_32 = S_MOV_B32 2147483647
-    ; SI-NEXT: [[S_AND_B32_:%[0-9]+]]:sreg_32 = S_AND_B32 [[COPY2]], [[S_MOV_B32_]], implicit-def dead $scc
-    ; SI-NEXT: [[REG_SEQUENCE:%[0-9]+]]:sreg_64 = REG_SEQUENCE [[COPY1]], %subreg.sub0, [[S_AND_B32_]], %subreg.sub1
+    ; SI-NEXT: [[COPY1:%[0-9]+]]:sreg_32_xm0 = COPY [[COPY]].sub1
+    ; SI-NEXT: [[S_AND_B32_:%[0-9]+]]:sreg_32 = S_AND_B32 [[COPY1]], [[S_MOV_B32_]], implicit-def dead $scc
+    ; SI-NEXT: [[COPY2:%[0-9]+]]:sreg_32_xm0 = COPY [[S_AND_B32_]]
+    ; SI-NEXT: [[COPY3:%[0-9]+]]:sreg_32_xexec_hi_and_sreg_32_xm0 = COPY [[COPY]].sub0
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/inst-select-fabs.mir:433]: weird that we gained this copy

File: llvm/utils/TableGen/GlobalISelEmitter.cpp [hunk 7]
```diff
@@ -288,11 +289,21 @@ static std::string getMangledRootDefName(StringRef DefOperandName) {
...
 
-static Expected<LLTCodeGen> getInstResultType(const TreePatternNode &Dst) {
+static Expected<LLTCodeGen> getInstResultType(const TreePatternNode &Dst,
+                                              const CodeGenTarget &Target) {
+  // While we allow more than one output (both implicit and explicit defs)
+  // below, we only expect one explicit def here.
+  assert(Dst.getOperator()->isSubClassOf("Instruction"));
+  CodeGenInstruction &InstInfo = Target.getInstruction(Dst.getOperator());
+  if (InstInfo.Operands.NumDefs != 1)
+    return failedImport("Dst pattern child is not having exactly one result");
```
Comment (arsenm) [inline llvm/utils/TableGen/GlobalISelEmitter.cpp:299]: Phrasing 
```suggestion
    return failedImport("Dst pattern child only supported with exactly one result");
```

**Review by arsenm — APPROVED** [review]

**Comment by bjope** [comment]
Landed as https://github.com/llvm/llvm-project/commit/da591d390e7f865c846d12dc5559875eca347c28
